### ISSUE TITLE
Socket leak in haproxy version 2.6.14

### ISSUE THREAD

**Issue by mtimm** [haproxy/haproxy#2249]
### Detailed Description of the Problem

We have a system where there are four haproxy instances residing on different virtual machines.  Two instances we consider "internal" to our product and the other two are for external connections.  For various reasons, the internal instances are running haproxy version 1.5.18 and the external instances are running haproxy 2.6.14.  As part of this setup we expose the stats from the external instances via the internal instances so that we can query the internal instances and get stats from the external ones using this configuration on the internal instances:

```
frontend haproxy_external_stats_fe
        bind :<redacted port number>
        option forwardfor
        default_backend haproxy_external_stats_be

backend haproxy_external_stats_be
        option httpchk GET /
        option forwardfor
        default-server inter 30s fall 3 rise 2
        server <redacted external instance1 ip> <redacted external instance1 ip>:<redacted port number> check
        server <redacted external instance2 ip> <redacted external instance2 ip>:<redacted port number> check
```

The external instances then have this configuration to serve up the stats:

```
global
  # HAProxy statistics unix socket
  stats socket /var/run/haproxy/haproxy-stats.sock mode 777 level user
  stats timeout 2m

frontend stats
  bind 127.0.0.1:<redacted port number>,<redacted ip address>:<redacted port number> thread 1 name stats
  stats enable
  stats uri /
```

With this configuration we see leaked sockets on the external instances related to the httpchk requests, from the lsof output against the haproxy pid:

```
haproxy 11800 haproxy   53u     sock                0,7      0t0 1904870076 protocol: TCP
haproxy 11800 haproxy   54u     sock                0,7      0t0 1903795796 protocol: TCP
haproxy 11800 haproxy   55u     sock                0,7      0t0 1904107382 protocol: TCP
haproxy 11800 haproxy   56u     sock                0,7      0t0 1904870080 protocol: TCP
haproxy 11800 haproxy   57u     sock                0,7      0t0 1905400204 protocol: TCP
haproxy 11800 haproxy   58u     sock                0,7      0t0 1904107380 protocol: TCP
haproxy 11800 haproxy   59u     sock                0,7      0t0 1905509562 protocol: TCP
haproxy 11800 haproxy   60u     sock                0,7      0t0 1905565681 protocol: TCP
haproxy 11800 haproxy   61u     sock                0,7      0t0 1905599026 protocol: TCP
haproxy 11800 haproxy   62u     sock                0,7      0t0 1905609791 protocol: TCP
haproxy 11800 haproxy   63u     sock                0,7      0t0 1905631248 protocol: TCP
haproxy 11800 haproxy   64u     sock                0,7      0t0 1905654366 protocol: TCP
haproxy 11800 haproxy   65u     sock                0,7      0t0 1905843599 protocol: TCP
haproxy 11800 haproxy   66u     sock                0,7      0t0 1906103448 protocol: TCP
haproxy 11800 haproxy   67u     sock                0,7      0t0 1906172268 protocol: TCP
haproxy 11800 haproxy   68u     sock                0,7      0t0 1906172269 protocol: TCP 
```

These sockets are not seen in netstat or ss output.  The `show fd` output for one of these sockets:

```
68 : st=0x001d22(cl HEOpI W:sRa R:sRa) ref=0 gid=1 tmask=0x1 umask=0x0 prmsk=0x0 pwmsk=0x0 owner=0x7f17780ef450 iocb=0x59aa50(sock_conn_iocb) back=0 cflg=0x801c0300 cerr=0 fam=ipv4 lport=18080 fe=stats mux=H1 ctx=0x7f1760160fb0 h1c.flg=0x107202 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 .task=0x7f17600b0650 .exp=<NEVER> h1s=0x7f17782b8c90 h1s.flg=0x104048 .sd.flg=0x2814a01 .req.state=MSG_DONE .res.state=MSG_DATA .meth=GET status=200 .sd.flg=0x02814a01 .sc.flg=0x00000280 .sc.app=0x7f175c1665c0 .subs=(nil) xprt=RAW 
```

And the show sess output for one:

```
0x7f175c1665c0: proto=tcpv4 src=<redacted ip address>:<redacted port> fe=stats be=<NONE> srv=<none> ts=00 epoch=0 age=4h44m calls=4 rate=0 cpu=0 lat=0 rq[f=4cc0c068h,i=32735,an=00h,rx=,wx=,ax=] rp[f=8804a860h,i=7494,an=00h,rx=,wx=,ax=] scf=[10,280h,fd=68] scb=[8,219h,fd=-1] exp= rc=0 c_exp=
```

Side note: we originally experienced this issue on 2.6.9 and attempted to upgrade to 2.6.14 but the same issue exists.

Side note 2: we compile from source using the Fedora Core 37 src rpm and link against our own version of ssl, this should not be related though since these endpoints do not have ssl.

### Expected Behavior

Haproxy should not leak sockets.

### Steps to Reproduce the Behavior

1. Run a httpchk from a haproxy 1.5.18 version (not sure if this is required) against a stats endpoint on a 2.6.14 haproxy version that is bound to a port.
2. Unknown what actually leads to the socket leak but it does not occur on lightly loaded systems for about 4 or 5 days.  On heavily loaded systems it occurs within the first hour.

### Do you have any idea what may have caused this?

This started when we upgraded the external instances from haproxy version 2.4.2 to 2.6.9.  We are not certain why the stats endpoint is leaking sockets though.

### What is your configuration?

```haproxy
global
        # The maximum number of connections accepted globally HAProxy
        maxconn         250000
        log 127.0.0.1   local0 info

        # Rate limit SSL connections to not overload the CPU
        maxsslrate 300

        # Enables processing SSL connection handshakes non-blocking
        ssl-mode-async

        # HAProxy statistics unix socket
        stats socket /var/run/haproxy/haproxy-stats.sock mode 777 level user
        stats timeout 2m

        # HAProxy daemon permissions and mode
        user haproxy
        group haproxy
        daemon

        # HAProxy 2.6 is fully multithreaded
        nbthread 52
        cpu-map auto:1/1-52 0-51

        # Default ciphers to use on SSL-enabled listening sockets.
        # For more information, see ciphers(1SSL).
        ssl-default-bind-ciphers ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA:!DSS
        ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets
        tune.ssl.default-dh-param 4096
        tune.ssl.cachesize 250000

        # Increase max buffer size from 16K to 32K to handle possible large requests.
        tune.bufsize 32768
# Define frontend and backend-level defaults.
defaults
        log     global              # Assume global logging defaults for each frontend/backend
        mode    http                # Use layer 7 HTTP mode for frontends and backends by default
        option  httplog             # Enable logging of HTTP request, session state and timers
        unique-id-format %ci:%cp_%fi:%fp_%Ts_%rt:%pid
        unique-id-header X-Request-Id
        option  dontlognull         # Don't log connections where no data was transferred
        option  forwardfor          # Enable insertion of the X-Forwarded-For header to requests sent to servers
        option  http-no-delay       # Favor low interactive delays over performance in HTTP
        option  http-server-close   # Enable HTTP connection closing on the server side
        maxconn 250000               # Maximum number of concurrent connections to allow per frontend/backend
        retries 3                   # The number of retries to perform on a server after a connection failure
        option  redispatch          # Enable session redistribution in case of connection failure
        timeout connect 5s          # The maximum time to wait for a connection attempt to a server to succeed
        timeout client  50s         # The maximum inactivity time on the client side
        timeout server  50s         # The maximum inactivity time on the server side
        timeout client-fin 50s      # Handle the situation that client disappears from the net
        timeout tunnel 1h           # Timeout to use with WebSocket and CONNECT

# other frontends and backends have been removed as they do not seem relevant to this issue

frontend stats
  bind 127.0.0.1:<redacted port>,<redacted ip address>:<redacted port> thread 1 name stats
  stats enable
  stats uri /
```

### Output of `haproxy -vv`

```plain
$ LD_LIBRARY_PATH="/opt/openssl/lib64" /usr/sbin/haproxy -vv
HAProxy version 2.6.14-5188364 2023/06/09 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.14.html
Running on: Linux 3.10.0-1160.92.1.el7.x86_64 #1 SMP Tue Jun 20 11:48:01 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY -LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -QUIC +RT +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=52).
Built with OpenSSL version : CompanySSL 1.1.1t.7.2.482
Running on OpenSSL version : CompanySSL 1.1.1t.7.2.482
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.23 2017-02-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 4.8.5 20150623 (Red Hat 4.8.5-28)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by wtarreau** [comment]
Thanks for your report. Do you remember what was the last version that did not exhibit this ?

Also what's the approximate leak rate, is it a few per hour, or more ? Do you think that once it starts to leak it speeds up, or maybe did you observe that they arrive in groups ?

If you happen to issue "show sess all" (or show sess 0xXXXX with its pointer), it would be really cool, because it would contain a lot more details than just "show sess". It will contain a few private info (source IP, frontend/backend names etc). You can turn on anon mode on the CLI to replace them with a hash if you want using "set anon on" before "show sess".

**Comment by mtimm** [comment]
The rate of the leak seems very consistent once it starts, it is about 170 sockets a day.  I'm not certain why but some systems do not seem to experience a leak right away while others do.

Here is a session output (it doesn't look like anon mode is available in haproxy version 2.6):

```
0x1ff91d0: [10/Aug/2023:18:15:20.833838] id=994 proto=tcpv4 source=<ip address redacted>:<port number redacted>
  flags=0x11c02, conn_retries=0, conn_exp=<NEVER> conn_et=0x200 srv_conn=(nil), pend_pos=(nil) waiting=0 epoch=0
  frontend=stats (id=5 mode=http), listener=stats (id=2) addr=<ip address redacted>:<port number redacted>
  backend=<NONE> (id=-1 mode=-)
  server=<NONE> (id=-1)
  task=0x7f175c1810e0 (state=0x00 nice=-32 calls=4 rate=0 exp=<NEVER> tmask=0x1 age=3d4h)
  txn=0x7f17504c5c30 flags=0x3000 meth=1 status=200 req.st=MSG_ERROR rsp.st=MSG_DATA req.f=0x44 rsp.f=0x0e
  scf=0x220d4b0 flags=0x00000280 state=CLO endp=CONN,0x7f17602cff10,0x02814a01 sub=0
      co0=0x2205720 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=LISTENER:0x207fcb0
      flags=0x801c0300 fd=51 fd.state=1d22 updt=0 fd.tmask=0x1
  scb=0x1ff95c0 flags=0x00000219 state=EST endp=APPCTX,0x1ff9630,0x01000002 sub=0
      app1=0x1ff9630 st0=2 st1=0 st2=0 applet=<STATS> tmask=0x1 nice=0 calls=6 rate=0 cpu=0 lat=0
  req=0x1ff91f0 (f=0x4cc0c068 an=0x0 pipe=0 tofwd=0 total=33)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x1ff91f8 data=0x21e51d0 o=33 p=33 i=32735 size=32768
      htx=0x21e51d0 flags=0x10 size=32720 data=33 used=2 wrap=NO extra=0
  res=0x1ff9250 (f=0x8804a860 an=0x0 pipe=0 tofwd=-1 total=148279)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x1ff9258 data=0x21ed1e0 o=25815 p=25815 i=6953 size=32768
      htx=0x21ed1e0 flags=0x0 size=32720 data=25815 used=2 wrap=NO extra=0
```

Here's an update from my side:

I did a packet capture between the two nodes and noticed that the response to the health check request was 150KB's in size (html response) but the client making the request (haproxy in this case) was sending a reset after the first one or two packets and the stats server just kept trying to send the rest of the data.  That kind of makes sense because the health check as configured is really only looking at the HTTP response code which is in the first packet.  The body is not even needed.

Based on this I initially changed the health check in the client to request the csv version of stats which was only 17KB in size and the leak seemed to stop.  Since the health check doesn't even need the body at all I have since changed it to request the HEAD method instead and I am monitoring now to see if we have any more leaked sockets.

**Comment by mtimm** [comment]
Sorry I just realized I didn't answer your very first question.

Originally we were running version 1.5.18 on both the client doing the health check against the stats endpoint and the server responding to the health check.  The leak was not seen.

Then we upgraded the server to version 2.6.9 and left the client at version 1.5.18.  This is where we noticed the leak.

We have since tested 2.6.14 on both the client and server and the leak is still seen in that version.

I could load previous versions in a staging environment to see what happens.  Since the leak is consistent regardless of the client, I would likely leave that alone for this test and just modify the server version.  The leak is very reminiscent of #1774 but in our case it is the stats server that seems to be experiencing the leak.

**Comment by wtarreau** [comment]
Thanks for all these precious details. I strongly suspect that this is related to a timing issue between the mux and the applet. Your trace above indeed shows the FD received an hangup and error. It should not have remained in this state. The thing is, I think that with a real server instead of the stats applet, the server is pushing to respond so this creates activity that allows to detect the problem, while here the applet doesn't move until the mux consumes its data, which does not happen. Most likely a wakeup and/or a flush of pending output is missing when this error arrives.

I'd be curious to know if "option abortonclose" in the frontend solves the issue.

I noted two weeks ago a few frozen connections on my home machine and I captured the "show sess all" output. The fd.state matches yours for some of them. I hadn't taken the time to analyze it yet because it was an old 2.7.5 and I wanted to update it first. But now I suspect we're having a long-time bug there. Ideally I'd prefer to wait for Christopher to be back to dig deeper in to mux-h1 as I'd rather avoid doing some subtle breakage. Given that the issue seems old, I guess it can wait a little bit as long as we find a workaround. Just let me know if that's OK for you.

**Comment by mtimm** [comment]
Yes we can wait, the workaround seems to be to limit the response by either requesting HEAD or requesting the csv stats.  Since we are not validating any part of the body as part of the health check, the HEAD request is sufficient for us.  Thanks for the responses @wtarreau!

**Comment by mtimm** [comment]
@wtarreau FYI:

> I'd be curious to know if "option abortonclose" in the frontend solves the issue.

Configuring this and doing a config check results in:

`[WARNING] 225/173337 (2296) : parsing [/etc/haproxy/haproxy.cfg:132] : 'abortonclose' ignored because frontend 'stats' has no backend capability.`

**Comment by wtarreau** [comment]
Ah sorry I forgot that it was only for the backend. So it will have no effect with the stats applet.

**Comment by mtimm** [comment]
No problem.  With such a complex piece of software I'm sure it is easy to forget small details like this.  Thank you again!

**Comment by wtarreau** [comment]
Indeed, a friend recently told me how many keywords we had. I thought hundreds, it's thousands!

**Comment by capflam** [comment]
Thanks @mtimm , I'm able to reproduce the issue quite easily. And I have a fix. It was not so obvious in fact because there were many ways to fix the issue. For now, I found an issue affecting all versions as far as 2.2. However, on recent versions (2.9 and 2.8), the result is not the same because of recent refactoring.

I can propose the following patch for the 2.7 and 2.6:

```diff
diff --git a/include/haproxy/sc_strm.h b/include/haproxy/sc_strm.h
index 82c583a605..01df4fa353 100644
--- a/include/haproxy/sc_strm.h
+++ b/include/haproxy/sc_strm.h
@@ -377,6 +377,8 @@ static inline int sc_is_send_allowed(const struct stconn *sc)
 
        if (oc->flags & CF_SHUTW)
                return 0;
+       if (oc->flags & CF_SHUTW_NOW)
+               return 1;
 
        return !sc_ep_test(sc, SE_FL_WAIT_DATA | SE_FL_WONT_CONSUME);
 }
diff --git a/src/stconn.c b/src/stconn.c
index 7b64566c87..bfacdc7df1 100644
--- a/src/stconn.c
+++ b/src/stconn.c
@@ -1014,7 +1014,7 @@ static void sc_app_chk_snd_applet(struct stconn *sc)
                return;
 
        /* we only wake the applet up if it was waiting for some data  and is ready to consume it */
-       if (!sc_ep_test(sc, SE_FL_WAIT_DATA) || sc_ep_test(sc, SE_FL_WONT_CONSUME))
+       if (!sc_ep_test(sc, SE_FL_WAIT_DATA|SE_FL_WONT_CONSUME) && !(oc->flags & CF_SHUTW_NOW))
                return;
 
        if (!tick_isset(oc->wex))
```

The applet must be notified when a shutdown is pending, even if it stated it will not consume data. On 2.4 and 2.2 the patch must be adapted. In addition to this patch, I must also fix the 2.9 and 2.7 to properly set the stream expiration date when outgoing data are blocked:

```diff
iff --git a/src/applet.c b/src/applet.c
index 7298860e5..f4c5c862e 100644
--- a/src/applet.c
+++ b/src/applet.c
@@ -467,8 +467,11 @@ struct task *task_run_applet(struct task *t, void *context, unsigned int state)

        if (channel_is_empty(sc_oc(sc)))
                sc_ep_report_send_activity(sc);
-       else
+       else {
                sc_ep_report_blocked_send(sc);
+               __sc_strm(sc)->task->expire = tick_first(__sc_strm(sc)->task->expire, sc_ep_snd_ex(sc));
+               task_queue(__sc_strm(sc)->task);
+       }

        /* measure the call rate and check for anomalies when too high */
        if (((b_size(sc_ib(sc)) && sc->flags & SC_FL_NEED_BUFF) || // asks for a buffer which is present
diff --git a/src/stconn.c b/src/stconn.c
index 8bffba73d..e4d876cc5 100644
--- a/src/stconn.c
+++ b/src/stconn.c
@@ -1133,7 +1133,6 @@ static void sc_notify(struct stconn *sc)
                (channel_is_empty(oc) && !oc->to_forward)))))) {
                task_wakeup(task, TASK_WOKEN_IO);
        }
-
        if (ic->flags & CF_READ_EVENT)
                sc->flags &= ~SC_FL_RCV_ONCE;
 }
@@ -1694,6 +1693,8 @@ static int sc_conn_send(struct stconn *sc)
                /* We couldn't send all of our data, let the mux know we'd like to send more */
                conn->mux->subscribe(sc, SUB_RETRY_SEND, &sc->wait_event);
                sc_ep_report_blocked_send(sc);
+               s->task->expire = tick_first(s->task->expire, sc_ep_snd_ex(sc));
+               task_queue(s->task);
        }

        return did_send;
```

I will wait a bit before merging these patches because I must evaluate another bug report related to this one. But it should do the trick in your case.

**Comment by mtimm** [comment]
@capflam Thank you very much for the root cause and patch suggestion.

**Comment by capflam** [comment]
FYI, several new fixes were merged related to this bug. It should be fixed in 2.9. Everything will be backported to 2.8 very soon because a new release is planned.

**Comment by capflam** [comment]
FYI, the 2.8.3  was released and the fixes were backported to 2.6
